camligorkem/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,164 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Optional;
+
+public class BinaryTree<T> {
+    private BinaryNode<T> root= null;
+
+    public BinaryTree(BinaryNode<T> root) {
+        this.root = root;
+    }
+
+    public BinaryNode<T> getRoot() {
+        return root;
+    }
+    public void setRoot(BinaryNode<T> root) {
+        this.root = root;
+    }
+
+    /**
+     * Insert new nodes to the tree, assumes root is NOT null
+     *  root should be set before it is used.
+     * @param root user should pass the root of the tree
+     * @param index to find the place to be inserted
+     * @param data new value to be added into tree
+     */
+    public void insertNode(BinaryNode<T> root, int index, T data){
+        BinaryNode<T> newNode = new BinaryNode<T>(data);
+        //check current node is empty
+        if(root == null) {
+            return;
+        }
+        int parentIndex = (index-1)/2;
+        if(parentIndex == 0){
+            if(root.getLeft()== null){
+                root.setLeft(newNode);
+            }else if(root.getRight() == null){
+                root.setRight(newNode);
+            }
+        }else{
+            BinaryNode<T> child;
+            if(parentIndex%2 ==1){
+                child= root.getLeft();
+            }else{
+                child = root.getRight();
+            }
+            insertNode(child, parentIndex, data);
+        }
+    }
+
+    /**
+     * Prints the binary tree in order (left- root- right)
+     * @param node prints the nodes including itself and its subtrees.
+     * If you pass root prints all the tree.
+     */
+    public void printTree(BinaryNode<T> node){
+        if(node != null){
+            printTree(node.getLeft());
+            System.out.print(node.getData()+ "\t");
+            printTree(node.getRight());
+        }
+    }
+
+    /**
+     * Insert new nodes to the tree assumes root is NOT null
+     *  root should be set before it is used.
+     *  Remark: When there is duplicate element returns the ancestor of the first found element
+     * @param root user should pass the root of the tree
+     * @param key the value of the node which its ancestors want to be found
+     * @return an array list of T which are ancestor values of the node, returns empty list if ancestor
+     * is the root, returns null if the tree is empty or node doesn't exist.
+     */
+    //Assumes unique value for each element
+    public ArrayList<T> getAncestor (BinaryNode<T> root,  T key){
+        if(root == null)
+            return null;
+        if(root.getData().equals(key)){    // key found - previous node is its ancestor
+            ArrayList<T> ancestorList = new ArrayList<>();
+            return ancestorList;
+        }
+        ArrayList<T> ancestorsL = getAncestor(root.getLeft(), key);
+        ArrayList<T> ancestorsR = getAncestor(root.getRight(), key);
+        if(ancestorsL!= null || ancestorsR!=null){ // key is present in the tree
```

Comment:
There is a potential optimisation: instead of calculating both children's subtrees before checking if any of them was successful you can calculate one and decide to calculate the other only if the first one was unsuccessful. Statistically it should improve your algorithm in half of the cases ;)

Reply:
I haven't think about that. It is actually a very good approach thanks a lot :)